File: src/main/scala/com/mikea/gae/rx/RxHttpResponse.java
package com.mikea.gae.rx;

import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

/**
 * @author mike.aizatsky@gmail.com
 */
public class RxHttpResponse {
    private final HttpServletResponse response;
    private boolean hasResponse;

    public RxHttpResponse(HttpServletResponse response) {
        this.response = response;
    }

    public boolean hasResponse() {
        return hasResponse;
    }

    public void sendRedirect(String url) throws IOException {
        hasResponse = true;
        response.sendRedirect(url);
    }

    public void sendError(int errorCode, String message) throws IOException {
        hasResponse = true;
        response.sendError(errorCode, message);

    }

    public void sendOk() throws IOException {
        hasResponse = true;
        response.sendError(200);
    }
}

File: src/main/scala/com/mikea/gae/rx/RxHttpResponse.scala
package com.mikea.gae.rx

import com.mikea.gae.rx.base.Transformer
import com.mikea.gae.rx.events.RxHttpRequest
import javax.servlet.http.HttpServletResponse

abstract class RxHttpResponse(val request : RxHttpRequest) {
  protected def render(httpResponse : HttpServletResponse): Unit

  private[rx] def render() : Unit = {
    render(request.httpResponse)
    request.processed = true
  }
}

case class RxHttpErrorResponse(errorCode: Int, message: String, override val request : RxHttpRequest) extends RxHttpResponse(request) {
  protected def render(httpResponse: HttpServletResponse) = httpResponse.sendError(errorCode, message)
}

case class RxHttpOkResponse(body : String, override val request : RxHttpRequest) extends RxHttpResponse(request) {
  def this(request : RxHttpRequest) = this(null, request)

  protected def render(httpResponse: HttpServletResponse) = {
    httpResponse.setStatus(200)
    if (body != null) {
      httpResponse.getOutputStream.print(body)
    }
  }
}
case class RxHttpRedirectResponse(url: String, override val request : RxHttpRequest) extends RxHttpResponse(request) {
  protected def render(httpResponse: HttpServletResponse) = httpResponse.sendRedirect(url)
}

object RxHttpResponse {
  def ok[Event <: RxHttpRequest](body : String) : Transformer[Event, RxHttpResponse] = Transformer.map((evt: Event) => new RxHttpOkResponse(body, evt))
  def ok[Event <: RxHttpRequest]() : Transformer[Event, RxHttpResponse] = Transformer.map((evt: Event) => new RxHttpOkResponse(evt))
}